### ISSUE TITLE
Export problems using typescript in an addon

### ISSUE THREAD

**Issue by cibernox** [typed-ember/ember-cli-typescript#12]
Seems that modules defined in `.ts` files don't get exported.

I have a reproduction example on this PR: https://github.com/cibernox/ember-native-dom-helpers/pull/16

The only changes in that PR, aside from installing the addon, is that I renamed `/test-support/helpers.js` -> `/test-support/helpers.ts`.

I see two unexpected things.

First of all, I see some disturbing error messages in the console that sound like errors but perhaps they are warnings because they don't cause the build to fail:

<img width="704" alt="screen shot 2017-02-27 at 17 08 41" src="https://cloud.githubusercontent.com/assets/265339/23371547/51ac3cc0-fd10-11e6-9e11-0e89b1624d9a.png">

The tests seems unable to resolve `ember-native-dom-helpers/test-support/helpers`. Inspecting `/assets/test-support.js`, there is no traces of `define('ember-native-dom-helpers/test-support/helpers')`, but the modules defined in `.js` files are there. 

The simples way to reproduce is clone the repo linked above, checkout the `migrate-to-typescript` branch and run `ember s`

**Comment by pixelhandler** [comment]
@locks do you have an example of an addon already using TypeScript we can follow as an example for test setup?

**Comment by pixelhandler** [comment]
@cibernox see https://github.com/emberwatch/ember-cli-typescript/issues/3#issuecomment-266247843 I'm curious if this addon is mainly for apps. 

@dustinfarris do you have TypeScript working in an addon?

**Comment by cibernox** [comment]
> exporting addon code to the addon name module (surprised this doesn't happen already)

@pixelhandler This seems exactly what I'm hitting. Seems that this addon is not ready for prime time :(

**Comment by dustinfarris** [comment]
I have not been able to use in an addon.  Was told not ready for that.

**Comment by rwjblue** [comment]
Awesome! Let's fix it!

**Comment by vitch** [comment]
I dug into this a bit...

The problem seems to be that the addon's folder isn't available in the tree that the `TypeScriptPreprocessor` acts on. 

The preprocessor is set up in `setupPreprocessorRegistry` via `registry.add('js', plugin);`... But it only seems to be invoked twice:

 1. With the merged contents from the `app` and `dummy/tests/app` directories; and
 2. With the contents of the `tests/` directory

It never seems to be called with the contents of the `addon` directory.

I have never used `setupPreprocessorRegistry` before - is this how it is meant to work? Is there a way to tell it you want to preprocess the addon's directory too? Or is the addon's directory merged into the app namespace at some point?

I'm happy to try and put together a PR if someone can tell me how this is meant to work...

**Comment by pixelhandler** [comment]
To isolate the issue for addons I tried creating an addon that uses the `treeForAddon` hook to compile the typescript but will need some time to figure it out. If anyone is interested in figuring it out, perhaps try working on https://gitlab.com/pixelhandler/ember-try-typescript

**Comment by pixelhandler** [comment]
@rwjblue I hear you may have talked to @krisselden about this issue today, thanks for thinking about this. Need help in the area of compiling from the `addon` directory, and likely `test-support` too. Seems that a build is unable to resolve modules from those directories.

**Comment by vitch** [comment]
After much chasing of ghosts I've managed to get to the bottom of this and it turns out that you can use `ember-cli-typescript` from an addon as is.

Here is an example:
https://github.com/vitch/addon-with-typescript

And an app which consumes that addon:
https://github.com/vitch/app-with-typescript

The thing that made it work for me was moving `ember-cli-typescript` from `devDependencies` to `dependencies`...

**Comment by pixelhandler** [comment]
@cibernox did you see this note from @vitch (above), now in README as well, https://github.com/emberwatch/ember-cli-typescript/blob/master/README.md#write-your-add-ons-in-typescript

**Comment by vitch** [comment]
I think @cibernox's problem was something different given [he already had it in dependencies](https://github.com/cibernox/ember-native-dom-helpers/pull/16/files).

It's possible that changes in `ember-cli` since `2.12.0-beta.1` or this addon since the issue was opened might have fixed it or there may have been something else at play in that case...

**Comment by pixelhandler** [comment]
@vitch the main goal @cibernox had was to use TypeScript with the `addon-test-support` directory of an addon, `ember-native-dom-helpers`. Curious if that just works by moving `ember-cli-typescript` to `dependencies`.

**Comment by rwjblue** [comment]
Because thats how preprocessors work. If an addon wants things in its own `addon` or `addon-test-support` folders to be transpiled when it is consumed, it must have a dependency on that preprocessor.  Any devDeps in an addon are _only_ used for the addon's own `dummy` test app.

**Comment by chriskrycho** [comment]
@rwjblue would you also need the `typescript` dependency in `dependencies` in that scenario? (As @vitch notes, @cibernox already has `ember-cli-typescript` in his `dependencies`: [see here](https://github.com/cibernox/ember-native-dom-helpers/pull/16/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R29).)

**Comment by cibernox** [comment]
I'll give this a try today or more likely tomorrow.

**Comment by toranb** [comment]
Any update @cibernox -would love to start using this with addons ;)

**Comment by toranb** [comment]
Disregard- got this working by moving `ember-cli-typescript` to dependencies (from devDependencies as mentioned above)

here is my tsconfig.json file w/ paths for others who might follow with the same quesiton

```js
{
  "compilerOptions": {
    "target": "es2015",
    "module": "es2015",
    "moduleResolution": "node",
    "noEmitOnError": false,
    "noEmit": true,
    "baseUrl": ".",
    "paths": {
      "q2-garnishment/*": ["addon/*", "app/*"],
      "ember-redux": ["node_modules/ember-redux/index.d.ts"]
    }
  },
  "include": [
    "**/*.ts"
  ]
}
```

And here is are the dependencies I'm using w/ ember-cli v2.10

```js
  "dependencies": {
    "ember-cli-babel": "^5.1.7",
    "ember-cli-htmlbars": "^1.0.10",
    "ember-cli-typescript": "git://github.com/toranb/ember-cli-typescript.git#master",
    "typescript": "2.3"
  }
```

**Comment by chriskrycho** [comment]
Resolved in #93.
